### ISSUE TITLE
restic --no_symlink option (FEATURE)

### ISSUE THREAD

**Issue by yhafri** [restic/restic#1062]
## Output of `restic version`

`any`

## Expected behavior

```
$ restic --no_symlink [...]
```

will tell _restic_ to totally ignore symlinks.

## Do you have any idea what may have caused this?

Symlinks can be a real nightmare. If we can have an option to ignore them during the backup process, it'll be very useful. Not only they can create problem during backup, but also during restore (permission denied).

**Comment by fd0** [comment]
Thanks for leaving an issue. I don't understand what your use case is, could you elaborate a bit? Why are symlinks "a real nightmare"?

What's the problem with symlinks during backup that you mention?

Can you describe how to reproduce "permission denied" errors when restoring symlinks?

**Comment by yhafri** [comment]
For example, on MacOS `/etc` is a symlink to `/private/etc`. If you try to restore it on let say `/tmp`, you get this:
```
ignoring error for /tmp/etc/xml/catalog: Lchown: lchown /tmp/etc/xml/catalog: operation not permitted
ignoring error for /tmp/etc/xml/catalog.old: Lchown: lchown /tmp/etc/xml/catalog.old: operation not permitted
ignoring error for /tmp/etc/xml/xml-core.xml: Lchown: lchown /tmp/etc/xml/xml-core.xml: operation not permitted
ignoring error for /tmp/etc/xml/xml-core.xml.old: Lchown: lchown /tmp/etc/xml/xml-core.xml.old: operation not permitted
ignoring error for /tmp/etc/zsh_command_not_found: Lchown: lchown /tmp/etc/zsh_command_not_found: operation not permitted
There were 2593 errors
```

In our case, we would like to ignore _symlinks_ from our backups definitely.

**Comment by fd0** [comment]
Hm, if `/etc` is a symlink to `/private/etc`, and you've made a backup that includes `/etc`, then restic will archive the symlink (not the target it points to). In this case, it will restore a symlink at `/tmp/etc`. But the error messages you've listed show that restic restored something within `/tmp/etc`, so in this case `/tmp/etc` must be a directory not a symlink. Did I miss something?

The permission denied error occurs apparently because `restore` was run using an unprivileged user account, so running `lchown` to e.g. `root` is not permitted. This error will occur every time restic restore something that originally belonged to a different user. This is not limited to symlinks, but will also happen for files, dirs, pipes etc. I fail to see the point why we need code that handles symlinks different from the rest. Is there something I don't see?

**Comment by yhafri** [comment]
@fd0 I've updated my answer above: `/tmp` is also a symlink to `/private/tmp`

<img width="539" alt="mac_os" src="https://user-images.githubusercontent.com/23661134/27507460-1f43804c-58d0-11e7-8d96-d2322dc19d19.png">

**Comment by fd0** [comment]
So, then just use `restic backup --exclude /etc --exclude /tmp --exclude /var`? Or, use `restic restore --exclued /tmp --exclude /var --exclude /etc`. Isn't that enough?

To be clear here: I'm trying to understand if we need new code, because adding code in general and special cases in particular increase the maintenance of the project as a whole. And I need to live with that code afterwards :)

**Comment by yhafri** [comment]
> So, then just use restic backup --exclude /etc --exclude /tmp --exclude /var? Or, use restic restore --exclued /tmp --exclude /var --exclude /etc. Isn't that enough?

Not enough IMHO, let state it otherwise: how can i tell `restic` to backup everything in directory *BUT* _symlinks_?

**Comment by fd0** [comment]
`restic backup --exclude-file $(find / -type l) /`

**Comment by yhafri** [comment]
> The permission denied error occurs apparently because restore was run using an unprivileged user account, so running lchown to e.g. root is not permitted. This error will occur every time restic restore something that originally belonged to a different user. This is not limited to symlinks, but will also happen for files, dirs, pipes etc. I fail to see the point why we need code that handles symlinks different from the rest. Is there something I don't see?

Why then let a backup to be performed as a normal user (in this case _me_), when you know that the restore will certainly fail because of permissions to be set (root) :-) ?
Moreover, I've `sudo` power on this machine and I'm part of `wheel` group.

**Comment by yhafri** [comment]
> restic backup --exclude-file $(find / -type l) /

Not portable. A simple option can let us even do it on Windows

**Comment by fd0** [comment]
> Why then let a backup to be performed as a normal user (in this case me), when you know that the restore will certainly fail because of permissions to be set (root) :-) ?

I don't understand this question, you ran restic, you instructed it to backup/restore. It tried, restored the files/symlinks and only complained that it is unable to set the ownership properly. For that, you could've started it as root...

**Comment by fd0** [comment]
> Not portable. A simple option can let us even do it on Windows

Now you're deviating from the original problem/use case you described. Moreover: Are symlinks on Windows a thing?

**Comment by yhafri** [comment]
> I don't understand this question, [...], you could've started it as root...

We never run any script as `root` in our case.

**Comment by fd0** [comment]
> We never run any script as root in our case.

So, what should restic do then? Call sudo on it's own? Where's the difference?

**Comment by yhafri** [comment]
> Now you're deviating from the original problem/use case you described. Moreover: Are symlinks on Windows a thing?

Not portable means even common tools (sed, find,etc.) don't expect the same arguments between unixes. And why we should rely on an external tool if this can be simply solved?

**Comment by fd0** [comment]
> And why we should rely on an external tool if this can be simply solved?

Because each and every feature increases code complexity, may contain bugs or not handle corner cases properly, must be tested and increases the overall maintenance burden. So I'd in order to implement something you need to convince me first that it solves a problem, because I need to live with and maintain the code afterwards.

**Comment by yhafri** [comment]
> So, what should restic do then? Call sudo on it's own? Where's the difference?

Simply fails if we're trying to backup files/dirs that don't belong to the current user.
Or at least give us a warning.

**Comment by yhafri** [comment]
> Because each and every feature increases code complexity, [...], because I need to live with and maintain the code afterwards.

Your call. I was just thinking that filtering symlinks in Go wasn't a big deal.
Thanks for your time.

**Comment by fd0** [comment]
The policy for errors is: Print the error and warn the user, but then try to continue. `backup` works just fine apparently (if you haven't seen any errors), but for restore restic followed the policy and let you know that it was unable to do something.

It seem to me we won't agree on this matter (which is fine), so I'm closing the issue for now. If you (or anybody else) have other arguments, then please add another comment. I'm not convinced (yet, at least).

Thanks for your time and explanations.

**Comment by yhafri** [comment]
> The policy for errors is: Print the error and warn the user, but then try to continue. backup works just fine apparently (if you haven't seen any errors), but for restore restic followed the policy and let you know that it was unable to do something.

You backup something and you can restore it. Why for?

**Comment by fd0** [comment]
> You backup something and you can restore it. Why for?

I'm sorry, I don't understand this question. Could you rephrase?

**Comment by yhafri** [comment]
[Symlinks on Windows](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365680(v=vs.85).aspx)

**Comment by yhafri** [comment]
You backup something and you _CAN'T_ restore it. Why for?

**Comment by fd0** [comment]
Thanks for the symlink link.

> You backup something and you CAN'T restore it. Why for?

That's a question you need to ask yourself: If you run `restore` without enough permissions to restore metadata such as ownership, why blame restic that it is unable to do something you did not grant it permission?

From your comments I read that you suggest the backup operation should fail (hard, as in: exit with an error code) when restic can anticipate that it will not be able to restore the files to backup. Please correct me if I'm wrong on that.

I can see several problems here:
 * `backup` and `restore` are two different operations. Let's say restic should backup `/etc/passwd`, which is a world-readable file owned by `root`. During backup, restic is run as user `nobody`. When `restore` is run as the same user `nobody`, restoring the file will fail because `nobody` cannot write to `/etc`. Even if the file is restored to a different directory, say `/tmp/restore/etc/passwd`, the ownership cannot be set. Should the backup operation fail? In your suggestion it should, but how could `restic` during `backup` know what the user will be during `restore`? The other way around: Run `backup` as root (so that it won't fail in your suggestion), but how should restic know which user account will be used for `restore`? When `restore` is run as user `nobody` restore fails with the same error. How could restic detect this situation?
 * It would be a bad user experience for users who e.g. backup `/usr` as the user `nobody` when restic fails with an error that it probably won't be able to restore something. The users will wonder because for the operation at hand (backup), the privileges granted to the restic process are perfectly fine, because everything in `/usr` can be read by the user `nobody`.

Did I miss anything?

**Comment by yhafri** [comment]
Perhaps we should split this ticket in two:
1. the hypothetical `no symlinks` feature. 
2. the `backup/restore` permissions 

I would like to continue here with a discussion on a portable way to exclude `symlinks` if you don't mind.

**Comment by fd0** [comment]
Good idea.

I'd like to summarize what I understood: You asked for an option to exclude all symlinks from being saved in a backup. For the reason you mentioned that symlinks can create problems during backup and restore, you only mentioned one problem for restore: Restic prints an error message when it is unable to set the correct ownership. When a lot of errors that are not relevant are printed it is easy to overlook really grave errors.

It feels to me as if there's an underlying issue, it may not even be about symlinks. I found the examples you included a bit confusing. Are you aware that the string `lchown` in the error messages is the name of a syscall used to set ownership on arbitrary items (files, dirs, symlinks, ...)?

**Comment by yhafri** [comment]
> I'd like to summarize what I understood: You asked for an option to exclude all symlinks from being saved in a backup. For a reason you mentioned that symlinks can create problems during backup and restore, you only mentioned one problem for restore: Restic prints an error message when it is unable to set the correct ownership. Is my summary accurate?

Correct ... this is exactly our concern. 
As you may know it by now, we use restic extensively under `Linux`, `MacOS`, and also `Windows`.

Having an option to _exclude symlinks in a portable way_ instead of relying on external tools (which may, or may not exist on different OSes) is a clear winner here.

Still, i see no reason for this new feature to affect anything in term of logic, as you already provide a good filtering mechanism (--exclude-file) based on filename patterns.

@fd0 Think of it as a complementary feature.

The permission issue described was just a side effect. I'll create a new ticket for it.

**Comment by fd0** [comment]
Ok, thanks for the confirmation. The permissions aside, could you please shed some light onto what problems you run into (or expect to run into) with symlinks during backup and restore, so I can better understand your situation/use case?
